File: email-service/app/circuit_breaker.py
import time
import asyncio
import logging
from typing import Callable, Any
from functools import wraps

logger = logging.getLogger(__name__)


class CircuitOpen(Exception):
    """Exception raised when circuit breaker is open"""
    pass


class CircuitBreaker:
    """
    Circuit Breaker pattern implementation
    States: CLOSED (normal), OPEN (failing), HALF_OPEN (testing recovery)
    """

    def __init__(self, fail_threshold=5, reset_timeout=60):
        self.fail_threshold = fail_threshold
        self.reset_timeout = reset_timeout
        self._failure_count = 0
        self._success_count = 0
        self._last_failure_time = None
        self._state = "CLOSED"
        logger.info(
            f"Circuit breaker initialized: threshold={fail_threshold}, "
            f"timeout={reset_timeout}s"
        )

    @property
    def state(self):
        """Get current circuit breaker state"""
        return self._state

    def record_success(self):
        """Record successful operation"""
        self._success_count += 1
        if self._state == "HALF_OPEN":
            # After successful test in half-open, close the circuit
            logger.info(
                "Circuit breaker recovery successful - closing circuit")
            self._failure_count = 0
            self._state = "CLOSED"
        elif self._state == "CLOSED":
            self._failure_count = max(0, self._failure_count - 1)

    def record_failure(self):
        """Record failed operation"""
        self._failure_count += 1
        self._last_failure_time = time.time()

        if self._state == "HALF_OPEN":
            # Failed during test, reopen circuit
            logger.warning("Circuit breaker test failed - reopening circuit")
            self._state = "OPEN"
        elif self._failure_count >= self.fail_threshold:
            logger.error(
                f"Circuit breaker threshold reached ({self._failure_count} failures) "
                f"- opening circuit"
            )
            self._state = "OPEN"

    def allow_request(self) -> bool:
        """Check if request is allowed based on circuit state"""
        if self._state == "CLOSED":
            return True

        if self._state == "OPEN":
            elapsed = time.time() - (self._last_failure_time or 0)
            if elapsed > self.reset_timeout:
                logger.info(
                    "Circuit breaker timeout reached - entering half-open state")
                self._state = "HALF_OPEN"
                return True
            return False

        # HALF_OPEN state - allow limited requests for testing
        return True

    def get_stats(self) -> dict:
        """Get circuit breaker statistics"""
        return {
            "state": self._state,
            "failure_count": self._failure_count,
            "success_count": self._success_count,
            "last_failure_time": self._last_failure_time,
        }

    def __call__(self, fn: Callable):
        """Decorator to wrap functions with circuit breaker"""
        @wraps(fn)
        async def wrapper(*args, **kwargs):
            if not self.allow_request():
                raise CircuitOpen(
                    f"Circuit breaker is OPEN. Service unavailable. "
                    f"Last failure: {self._last_failure_time}"
                )
            try:
                result = await fn(*args, **kwargs)
                self.record_success()
                return result
            except Exception as e:
                self.record_failure()
                logger.error(f"Circuit breaker caught exception: {e}")
                raise
        return wrapper


class RetryPolicy:
    """
    Retry policy with exponential backoff
    """

    def __init__(self, max_retries=3, base_delay=1, max_delay=30):
        self.max_retries = max_retries
        self.base_delay = base_delay
        self.max_delay = max_delay

    def calculate_delay(self, attempt: int) -> float:
        """Calculate exponential backoff delay"""
        delay = min(self.base_delay * (2 ** attempt), self.max_delay)
        return delay

    def __call__(self, fn: Callable):
        """Decorator to add retry logic to functions"""
        @wraps(fn)
        async def wrapper(*args, **kwargs):
            last_exception = None

            for attempt in range(self.max_retries + 1):
                try:
                    return await fn(*args, **kwargs)
                except CircuitOpen:
                    # Don't retry if circuit is open
                    raise
                except Exception as e:
                    last_exception = e
                    if attempt < self.max_retries:
                        delay = self.calculate_delay(attempt)
                        logger.warning(
                            f"Retry attempt {attempt + 1}/{self.max_retries} "
                            f"after {delay}s due to: {e}"
                        )
                        await asyncio.sleep(delay)
                    else:
                        logger.error(
                            f"All retry attempts exhausted ({self.max_retries}). "
                            f"Last error: {e}"
                        )

            raise last_exception
        return wrapper

File: push-service/src/circuit-breaker.ts
import { Injectable, Logger } from '@nestjs/common';

export enum CircuitState {
  CLOSED = 'CLOSED',
  OPEN = 'OPEN',
  HALF_OPEN = 'HALF_OPEN',
}

export class CircuitOpenError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CircuitOpenError';
  }
}

@Injectable()
export class CircuitBreaker {
  private readonly logger = new Logger(CircuitBreaker.name);
  private state: CircuitState = CircuitState.CLOSED;
  private failureCount = 0;
  private successCount = 0;
  private lastFailureTime: number | null = null;
  private readonly failThreshold: number;
  private readonly resetTimeout: number;

  constructor(failThreshold = 5, resetTimeout = 60000) {
    this.failThreshold = failThreshold;
    this.resetTimeout = resetTimeout;
    this.logger.log(
      `Circuit breaker initialized: threshold=${failThreshold}, timeout=${resetTimeout}ms`,
    );
  }

  getState(): CircuitState {
    return this.state;
  }

  recordSuccess(): void {
    this.successCount++;
    if (this.state === CircuitState.HALF_OPEN) {
      this.logger.log('Circuit breaker recovery successful - closing circuit');
      this.failureCount = 0;
      this.state = CircuitState.CLOSED;
    } else if (this.state === CircuitState.CLOSED) {
      this.failureCount = Math.max(0, this.failureCount - 1);
    }
  }

  recordFailure(): void {
    this.failureCount++;
    this.lastFailureTime = Date.now();

    if (this.state === CircuitState.HALF_OPEN) {
      this.logger.warn('Circuit breaker test failed - reopening circuit');
      this.state = CircuitState.OPEN;
    } else if (this.failureCount >= this.failThreshold) {
      this.logger.error(
        `Circuit breaker threshold reached (${this.failureCount} failures) - opening circuit`,
      );
      this.state = CircuitState.OPEN;
    }
  }

  allowRequest(): boolean {
    if (this.state === CircuitState.CLOSED) {
      return true;
    }

    if (this.state === CircuitState.OPEN) {
      const elapsed = Date.now() - (this.lastFailureTime || 0);
      if (elapsed > this.resetTimeout) {
        this.logger.log(
          'Circuit breaker timeout reached - entering half-open state',
        );
        this.state = CircuitState.HALF_OPEN;
        return true;
      }
      return false;
    }

    // HALF_OPEN state - allow limited requests for testing
    return true;
  }

  getStats() {
    return {
      state: this.state,
      failure_count: this.failureCount,
      success_count: this.successCount,
      last_failure_time: this.lastFailureTime,
    };
  }

  async execute<T>(fn: () => Promise<T>): Promise<T> {
    if (!this.allowRequest()) {
      throw new CircuitOpenError(
        `Circuit breaker is OPEN. Service unavailable. Last failure: ${this.lastFailureTime}`,
      );
    }

    try {
      const result = await fn();
      this.recordSuccess();
      return result;
    } catch (error) {
      this.recordFailure();
      this.logger.error(`Circuit breaker caught exception: ${error.message}`);
      throw error;
    }
  }
}

export class RetryPolicy {
  private readonly logger = new Logger(RetryPolicy.name);
  private readonly maxRetries: number;
  private readonly baseDelay: number;
  private readonly maxDelay: number;

  constructor(maxRetries = 3, baseDelay = 1000, maxDelay = 30000) {
    this.maxRetries = maxRetries;
    this.baseDelay = baseDelay;
    this.maxDelay = maxDelay;
  }

  calculateDelay(attempt: number): number {
    const delay = Math.min(this.baseDelay * Math.pow(2, attempt), this.maxDelay);
    return delay;
  }

  async execute<T>(fn: () => Promise<T>): Promise<T> {
    let lastError: Error;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      try {
        return await fn();
      } catch (error) {
        lastError = error;

        if (error instanceof CircuitOpenError) {
          // Don't retry if circuit is open
          throw error;
        }

        if (attempt < this.maxRetries) {
          const delay = this.calculateDelay(attempt);
          this.logger.warn(
            `Retry attempt ${attempt + 1}/${this.maxRetries} after ${delay}ms due to: ${error.message}`,
          );
          await this.sleep(delay);
        } else {
          this.logger.error(
            `All retry attempts exhausted (${this.maxRetries}). Last error: ${error.message}`,
          );
        }
      }
    }

    throw lastError;
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
